media_utils/eme_puppeteer.py

```diff
@@ -0,0 +1,138 @@
+                       .getInterface(Components.interfaces.nsIWebNavigation)\
+                       .QueryInterface(Components.interfaces.nsIDocShellTreeItem)\
+                       .rootTreeItem\
+                       .QueryInterface(Components.interfaces.nsIInterfaceRequestor)\
+                       .getInterface(Components.interfaces.nsIDOMWindow);\
+var tabbrowser = mainWindow.gBrowser;\
+for (var i=0; i < tabbrowser.browsers.length; ++i) {\
+  var b = tabbrowser.getBrowserAtIndex(i);\
+  var media = b.contentDocumentAsCPOW.getElementsByTagName('video');\
+  for (var j=0; j < media.length; ++j) {\
+     var ms = media[j].mozMediaSourceObject;\
+     if (ms) {\
+       debugLines = ms.mozDebugReaderData.split(\"\\n\");\
+       return debugLines;\
+     }\
+  }\
+}"
+
+def playback_started(eme):
+    try:
+        current_time = eme.player_current_time
+        return current_time > 0
+    except BaseException as e:
+        print ('Got exception %s' % e)
+        return False
+
+class EMEException(Exception):
+    pass
+
+# Debug info looks something like:
+
+#	Dumping data for reader d10c000:
+#		Dumping Audio Track Decoders: - mLastAudioTime: 9.386666
+#			Reader 0: d2df400 ranges=[(0.000000, 16.042666)] active=true size=196456
+#		Dumping Video Track Decoders - mLastVideoTime: 8.466791
+#			Reader 2: 13f54400 ranges=[(8.008000, 16.016000)] active=true size=3741917
+#			Reader 1: d2edc00 ranges=[(4.004000, 8.008000)] active=false size=1088692
+#			Reader 0: d2eac00 ranges=[(2.002000, 4.004000)] active=false size=242367
+#
+# One simple hack: There should be two active readers at any one time. If there are not exactly two,
+# something is horribly wrong.
+
+active_re = re.compile("active=true", re.DOTALL)
+
+def playback_done(eme):
+    time_left = eme.player_duration - eme.player_current_time
+    if abs(time_left) < 2.0:
+        return True
+
+    # Now, parse out playing info. If there is no active audio stream or no active video stream, throw
+    # an exception.
+
+    debug_lines = eme.player_get_debug_lines()
+    if debug_lines == None:
+        return True
+
+    count = 0
+    for line in debug_lines:
+        match = active_re.findall(line)
+        if len(match) > 0:
+            count = count + 1
+
+    if count == 2:
+        return False
+
+    raise EMEException('Did not find exactly one audio and one video active reader - url %s at %s' % (eme.url, eme.player_current_time))
+
+class EMEPuppeteer:
+    """
+    Wrapper around an EME media-player element
+    """
+
+    def __init__(self, marionette, url):
+        self.marionette = marionette
+        self.url = url
+        wait = Wait(self.marionette, timeout=30)
+        with self.marionette.using_context('content'):
+            self.marionette.navigate(self.url)
+            wait.until(expected.element_present(By.TAG_NAME, 'video'))
+            self.video = self.marionette.find_elements(By.TAG_NAME, 'video')[0]
```

Comment:
This can be `self.marionette.find_element(By.TAG_NAME, 'video')` -- it just returns the first element it finds.